spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java

```diff
@@ -550,17 +533,7 @@ protected void doShutdown() {
 	private void actualShutDown() {
 		Assert.state(getTaskExecutor() != null, "Cannot shut down if not initialized");
 		logger.debug("Shutting down");
-		for (DefaultConsumer consumer : this.consumers) {
-			try {
-				if (this.logger.isDebugEnabled()) {
-					this.logger.debug("Canceling " + consumer);
-				}
-				consumer.getChannel().basicCancel(consumer.getConsumerTag());
-			}
-			catch (IOException e) {
-				this.logger.error("Cancel Error", e);
-			}
-		}
+		new LinkedList<>(this.consumers).forEach(this::cancelConsumer);
```

Comment:
I think we need to avoid `ConcurrentModificationException`, because `cancelConsumer()` performs `this.consumers.remove()`